File: architecture-skills/design-patterns/examples/command/command.py
"""Command pattern in Python.

Text editor with undo: each command mutates the editor and knows how to
reverse itself; a history stack replays undo in LIFO order.

Note: for one-shot commands without undo, a plain function or closure is
the Pythonic form (shown at the bottom); classes earn their keep once
commands carry undo state.
"""

from __future__ import annotations


class Editor:
    """Receiver: the object commands operate on."""

    def __init__(self) -> None:
        self.text = ""


class InsertText:
    def __init__(self, editor: Editor, text: str) -> None:
        self._editor = editor
        self._text = text

    def execute(self) -> None:
        self._editor.text += self._text

    def undo(self) -> None:
        self._editor.text = self._editor.text[: -len(self._text)]


class DeleteLast:
    def __init__(self, editor: Editor, count: int) -> None:
        self._editor = editor
        self._count = count
        self._deleted = ""  # captured on execute so undo can restore it

    def execute(self) -> None:
        self._deleted = self._editor.text[-self._count:]
        self._editor.text = self._editor.text[: -self._count]

    def undo(self) -> None:
        self._editor.text += self._deleted


class History:
    """Invoker: runs commands and tracks them for undo."""

    def __init__(self) -> None:
        self._undo_stack: list[InsertText | DeleteLast] = []

    def execute(self, command: InsertText | DeleteLast) -> None:
        command.execute()
        self._undo_stack.append(command)

    def undo(self) -> None:
        if self._undo_stack:
            self._undo_stack.pop().undo()


if __name__ == "__main__":
    editor = Editor()
    history = History()

    history.execute(InsertText(editor, "Hello"))
    history.execute(InsertText(editor, ", world!"))
    history.execute(DeleteLast(editor, 1))
    print(f"after edits: {editor.text!r}")

    history.undo()
    history.undo()
    print(f"after 2 undos: {editor.text!r}")

    # Closure form: fine when no undo state is needed.
    shout = lambda: print(editor.text.upper())  # noqa: E731
    shout()

File: architecture-skills/design-patterns/examples/command/command.go
// Command pattern in Go.
//
// Text editor with undo: a small Command interface with Execute/Undo
// and a history slice as the invoker. For fire-and-forget commands
// without undo, a plain func() value is the idiomatic Go form.

package main

import "fmt"

// Editor is the receiver.
type Editor struct {
	Text string
}

type Command interface {
	Execute()
	Undo()
}

type InsertText struct {
	editor *Editor
	text   string
}

func (c *InsertText) Execute() { c.editor.Text += c.text }
func (c *InsertText) Undo()    { c.editor.Text = c.editor.Text[:len(c.editor.Text)-len(c.text)] }

type DeleteLast struct {
	editor  *Editor
	count   int
	deleted string // captured on Execute so Undo can restore it
}

func (c *DeleteLast) Execute() {
	split := len(c.editor.Text) - c.count
	c.deleted = c.editor.Text[split:]
	c.editor.Text = c.editor.Text[:split]
}

func (c *DeleteLast) Undo() { c.editor.Text += c.deleted }

// History is the invoker: it runs commands and tracks them for undo.
type History struct {
	undoStack []Command
}

func (h *History) Execute(c Command) {
	c.Execute()
	h.undoStack = append(h.undoStack, c)
}

func (h *History) Undo() {
	if n := len(h.undoStack); n > 0 {
		h.undoStack[n-1].Undo()
		h.undoStack = h.undoStack[:n-1]
	}
}

func main() {
	editor := &Editor{}
	history := &History{}

	history.Execute(&InsertText{editor, "Hello"})
	history.Execute(&InsertText{editor, ", world!"})
	history.Execute(&DeleteLast{editor: editor, count: 1})
	fmt.Printf("after edits: %q\n", editor.Text)

	history.Undo()
	history.Undo()
	fmt.Printf("after 2 undos: %q\n", editor.Text)
}
